core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala

```diff
@@ -189,4 +198,56 @@ class ServerShutdownTest extends ZooKeeperTestHarness {
     server.awaitShutdown()
     server.shutdown()
   }
+
+  // Verify that if controller is in the midst of processing a request, shutdown completes
+  // without waiting for request timeout.
+  @Test
+  def testControllerShutdownDuringSend(): Unit = {
+    val securityProtocol = SecurityProtocol.PLAINTEXT
+    val listenerName = ListenerName.forSecurityProtocol(securityProtocol)
+
+    val controllerId = 2
+    val metrics = new Metrics
+    val executor = Executors.newSingleThreadExecutor
+    var serverSocket: ServerSocket = null
+    var controllerChannelManager: ControllerChannelManager = null
+
+    try {
+      // Set up a server to accept connections, but not send any responses to controller requests
+      serverSocket = new ServerSocket(0)
+      val connectionFuture = executor.submit(new Runnable {
+        override def run(): Unit = serverSocket.accept()
+      })
+
+      // Start a ControllerChannelManager
+      val brokers = Seq(new Broker(1, "localhost", serverSocket.getLocalPort, listenerName, securityProtocol))
+      val controllerConfig = KafkaConfig.fromProps(TestUtils.createBrokerConfig(controllerId, zkConnect))
+      val controllerContext = new ControllerContext
+      controllerContext.liveBrokers = brokers.toSet
+      controllerChannelManager = new ControllerChannelManager(controllerContext, controllerConfig, Time.SYSTEM,
+        metrics, new StateChangeLogger(controllerId, inControllerContext = true, None))
+      controllerChannelManager.startup()
+
+      // Initiate a sendRequest and wait for connection to be established
+      val requestBuilder = new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion,
+        controllerId, 1, Map.empty.asJava, brokers.map(_.node(listenerName)).toSet.asJava)
+      controllerChannelManager.sendRequest(1, ApiKeys.LEADER_AND_ISR, requestBuilder)
+      connectionFuture.get(10, TimeUnit.SECONDS)
+      Thread.sleep(10)   // leave a little time to ensure send has been initiated
```

Comment:
I guess there may not be a non-intrusive way to accomplish this more reliably. Perhaps we could check the size of the message queue?

Reply:
I was being lazy here, updated the test to wait for the server to receive one byte.